vnext/ReactUWP/Views/TouchEventHandler.cpp

```diff
@@ -400,19 +400,36 @@ const char *TouchEventHandler::GetTouchEventTypeName(TouchEventType eventType) n
 bool TouchEventHandler::TagFromOriginalSource(
     const winrt::PointerRoutedEventArgs &args,
     int64_t *pTag,
-    xaml::UIElement *pSourceElement) {
+    winrt::UIElement *pSourceElement) {
   assert(pTag != nullptr);
   assert(pSourceElement != nullptr);
 
   // Find the React element that triggered the input event
-  xaml::UIElement sourceElement = args.OriginalSource().try_as<xaml::UIElement>();
+  winrt::UIElement sourceElement = args.OriginalSource().try_as<winrt::UIElement>();
   winrt::IPropertyValue tag(nullptr);
+
   while (sourceElement) {
-    tag = sourceElement.GetValue(xaml::FrameworkElement::TagProperty()).try_as<winrt::IPropertyValue>();
+    tag = sourceElement.GetValue(winrt::FrameworkElement::TagProperty()).try_as<winrt::IPropertyValue>();
     if (tag) {
+      // If a TextBlock was the UIElement event source, perform a more accurate hit test,
+      // searching for the tag of the nested Run/Span XAML elements that the user actually clicked.
+      // This is to support nested <Text> elements in React.
+      // Nested React <Text> elements get translated into nested XAML <Span> elements,
+      // while the content of the <Text> becomes a list of XAML <Run> elements.
+      if (const auto textBlock = sourceElement.try_as<winrt::Controls::TextBlock>()) {
+        const auto pointerPos = args.GetCurrentPoint(textBlock).RawPosition();
+        const auto inlines = textBlock.Inlines().GetView();
+
+        auto finerTag = TestHit(inlines, pointerPos);
```

Comment:
>auto finerTag = TestHit(inlines, pointerPos); [](start = 8, length = 45)

 We have an optimization where if the Text component only has one run , we avoid using the inline collection and set the Text property instead.  This allows XAML to go down a "fast path" for text.  

For example:
<Text>abc</Text>
should go down this path.

See:  RawTextViewManager::UpdateProperty() - there's a case where we check the parent for 1 child and set TextBlock.Text in this case.

In this case you may get an empty InlineCollection for the TextBlock.  Can you test this scenario?

If this turns out to be the case, you could just use the TextBlock tag instead of calling TestHit.

Reply:
OK I double-checked on this.  XAML has the smarts to stay in fast path even if you query for the Inlines property, and Inlines will not be empty.  You're good to go.

---
In reply to: [435664020](https://github.com/microsoft/react-native-windows/pull/5125#discussion_r435664020) [](ancestors = 435664020)

Reply:
I've debugged this scenario (\<Text> React Native \</Text>), and while we do set the Text() property on the TextBlock, and don't set Inlines() manually, the InlineCollection is not empty/null, and still contains one Run. 
(Also, why would the optimization be impeded, if I query for Inlines - would it get dynamically populated if I did, and then XAML would not look at the Text() property?)
Interestingly, in this scenario, GetCharacterRect() calls return (0.0) which is not true, and the function falls back to the tag of the TextBlock it found on line 414. I suppose I might as well, skip TestHit in this case, and check if the Text() property is set (we clear it if there're more children).

Reply:
Sorry if you didn't see my closed comment notes.  Summary is your code is fine. ...

If you aren't already using CodeFlow for responding to PR comments, I highly recommend it.  See:  [https://chrome.google.com/webstore/detail/codeflow/aphnoipocoffpdafmiidfmaiadhilelm](https://chrome.google.com/webstore/detail/codeflow/aphnoipocoffpdafmiidfmaiadhilelm)


---
In reply to: [435752667](https://github.com/microsoft/react-native-windows/pull/5125#discussion_r435752667) [](ancestors = 435752667)

Reply:
I'll check it out, thanks!